hermes/bin/src/packaging/metadata.rs

```diff
@@ -0,0 +1,71 @@
+//! A generalized metadata object which is used an specified by the Hermes application
+//! package or Hermes WASM module package.
+
+use std::{
+    fmt::{Debug, Display},
+    io::Read,
+};
+
+use chrono::{DateTime, Utc};
+
+use crate::packaging::schema_validation::SchemaValidator;
+
+/// Metadata object.
+pub(crate) struct Metadata<T> {
+    /// metadata JSON object.
+    json: serde_json::Map<String, serde_json::Value>,
+    /// `T` type marker.
+    _phantom: std::marker::PhantomData<T>,
+}
+
+impl<T> PartialEq for Metadata<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.json == other.json
+    }
+}
+impl<T> Eq for Metadata<T> {}
+impl<T> Display for Metadata<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.json.fmt(f)
+    }
+}
+impl<T> Debug for Metadata<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.json.fmt(f)
+    }
+}
+
+/// Traits defines a specific metadata schema specified for some type.
+pub(crate) trait MetadataSchema {
+    /// Metadata schema JSON string definition.
+    const METADATA_SCHEMA: &'static str;
+}
+
+impl<T: MetadataSchema> Metadata<T> {
+    /// Create `Metadata` from reader.
+    pub(crate) fn from_reader(reader: impl Read) -> anyhow::Result<Self> {
+        let schema_validator = SchemaValidator::from_str(T::METADATA_SCHEMA)?;
+        let json = schema_validator.deserialize_and_validate(reader)?;
+        Ok(Self {
+            json,
+            _phantom: std::marker::PhantomData,
```

Comment:
why do we need phantom data here?

Reply:
```rust
pub(crate) struct Metadata<T> {
    /// metadata JSON object.
    json: serde_json::Map<String, serde_json::Value>,
    /// `T` type marker.
    _phantom: std::marker::PhantomData<T>,
}
```

The `Metadata` itself holds just a general purpose ` json: serde_json::Map<String, serde_json::Value>,` as a field.
I've added generic `T`  so in the case where application has their own metadata object and wasm module has their own metadata object, to not mess-up, have a type safety and distinguish these two different metadata objects.

P.S. Important to add that I've also added the following trait 
```rust
pub(crate) trait MetadataSchema {
    /// Metadata schema JSON string definition.
    const METADATA_SCHEMA: &'static str;
}
```
which should be implement by the type `T` to instantiate a `Metadata<T>` value.
This trait is used to statically specify a json schema for the metadata object.